gcs_light_client.mediawiki

```diff
@@ -0,0 +1,992 @@
+through a PRF. This creates a set of uniformly distributed values. If we then
+sort each of these values, the ''delta'' between each of the values closely
+resembles a ''Geometric Distribution''. We'll again leverage this fact to use
+Golomb-Rice coding to compresses our set by only encoding the ''delta'' between
+two successive elements in the set.
+
+Golomb-Rice coded sets take two parameters: 
+* <code>N</code> the number of items to be inserted into the set
+* <code>P</code> a value which is computed as <code>1/fp</code> where <code>fp</code> is the desired false positive rate. 
+    
+<code>P</code> can also be interpreted as the parameter to our Geometric
+Distribution.  Intuitively, to achieve a false positive rate of 1/32 (1/2^5),
+in a series of queries of items which ''aren't'' in the set, we expect to
+receive a "NO" (false) 32 times, before getting a "YES" (true, our false
+positive). Once again, <code>P</code> MUST be a power of two.
+
+Given these two parameters, we can now construct our set.
+
+==== Set Construction ====
+
+Set construction takes three parameters: <code>N</code>, <code>P</code> and
+<code>L</code>
+* where <code>L</code> is a list of the ''raw'' items we wish to insert into the set
+* the type of <code>L</code> is assumed to be of <code>[]byte</code>
+
+NOTE: <code>P</code> ''must'' be a power of two as we target the specialized case of
+Golomb coding: Golomb-Rice coding.
+
+Using <code>N</code> and <code>P</code> we compute <code>F = N * P</code>
+<code>F</code> constricts the range of the hashed values accordingly in order to
+achieve our desired false positive rate.
+
+In addition, to help optimize the algorithm, we use a fast range algorithm[7],
+multiplying the hashed value by F and taking only the top 64 bits. This fairly
+distributes the values over F without division and can be done with fewer cycles
+cycles on CPUs that have 128-bit registers.
+
+The following routine computes the ''uncompressed'' set given the parameters
+above:
+<pre>
+hashed_set_construct(N, P, raw_items, k): -> []uint64:
+    let F = N * P
+
+    let set_items = []
+    for item in raw_items:
+        let set_value = (siphash(k, item) * F) >> 64
+        set_items.append(set_value)
+
+    set_items.sort()
+
+    set_items
+</pre>
+
+Using the routine above, we can transform our set of (possibly heterogeneous
+items) in to a list of uniformly distributed values. As a final step, these
+values are then sorted.
+
+==== Set Compression ====
+
+Once the set of hashed items has been constructed (and sorted) we then use
+Golomb-Rice coding to ''compress'' the set by encoding the ''delta'' value
+between each successive element within the set. As these values are uniformly
+distributed, the deltas between these values will be Geometrically Distributed,
+meaning that Golomb-Rice coding will be optimal for this use-case [6].
+
+The following routine describes the compression process:
+<pre>
+gcs_compress(sorted_set, fp) -> []byte:
```

Comment:
Wait, sorted_set? No where else does anything reference that the set must be sorted? Sorted in what order?

Reply:
The `hashed_set_construct` function above this returns a sorted set. Update to point this out, and make it explicit in the section above `gcs_compress`.